### ISSUE TITLE
Including the key in the mapper function 

### ISSUE THREAD

**Issue by mjrussell** [HeyImAlex/reselect-map#4]
Great utility library! Ran into this exact scenario and was about to implement it myself until I stumbled across this.

One thing that would be great would be to add support for getting the key in the mapper function in `createObjectSelector`. For instance I might have an example like:

```js
import { createObjectSelector } from 'reselect-map'

const exampleState = {
  numbers: { a: 1, b: 2 },
}

const mul = createObjectSelector(
  state => state.numbers,
  state => state.multiplier,
  (number, key, multiplier) => `${key}-${number * multiplier}`
)

console.log(mul(exampleState)) // { a: 'a-5', b: 'b-10' }
```

Lodash and other utility libraries typically provide the key also, typically as the second argument. Seems like it would be easy to change https://github.com/HeyImAlex/reselect-map/blob/master/src/memoize.js#L80 from 
```js
const result = fn(value, ...args)
```
to
```js
const result = fn(value, key, ...args)
```

Obviously this is a breaking change and not every mapper would necessarily need the key, so thought I'd see what approach you'd take before submitting a PR

**Comment by HeyImAlex** [comment]
Hm, I would be hesitant to add this only because it's a breaking change, but tentatively I think you're right that it's a better api. I'm about to head home so I'll mull over this tonight.

**Comment by mjrussell** [comment]
I supposed you could throw it as the last argument to the mapper and it wouldn't be breaking. Seems a little clunky back there but better than nothing!

**Comment by HeyImAlex** [comment]
I'm not completely opposed to a breaking change, and it looks like a good choice. I'm going to sit on this issue for a little while though, until I have time to think about it some more.

**Comment by LiborOl** [comment]
Hi @mjrussell, can use use Array selector instead of Object selector? Something like this.

`import { createArraySelector } from 'reselect-map'

const exampleState = {
  numbers: [ {key: 'a', value: 1}, {key: 'b', value: '2' }],
}

const mul = createArraySelector(
  state => state.numbers,
  state => state.multiplier,
  (number, multiplier) => `${number.key}-${number.value * multiplier}`
)

console.log(mul(exampleState)) // [ a: 'a-5', b: 'b-10' ]`

It could solve your problem.  

Anyway, I agree the API with key on object selector would be nicer. But if you do this change, to be consistent, there should be also index on array selector API ("Why is the key on object selector but not the index on array selector?"). But for array selector its performance killer (you will have to reevaluate array items because of index change - this will break the main benefit of array selector).

**Comment by mjrussell** [comment]
@LiborOl if I was going to walk the data set to create an array as you describe, I might as well just `_.map` across the object and put the key inside the object values. However the whole point of this library is to prevent unnecessarily walking through the map each time.

>  But for array selector its performance killer (you will have to reevaluate array items because of index change - this will break the main benefit of array selector).

This would only be true if you are not inserting at the end of your array.

**Comment by LiborOl** [comment]
I think, it wouldn't be so easy. The _.map will create new objects so all the items will be reevaluated. You should reuse the old ones. Or I am missing something?

> This would only be true if you are not inserting at the end of your array.

Of course just for items with changed indices (and not only inserting but also deleting). E.g. we are inserting in one big array at the beginning (the newest items on the top - not so rare I think). So we will not be happy at all with the index in the API :-)

**Comment by mjrussell** [comment]
> I think, it wouldn't be so easy. The _.map will create new objects so all the items will be reevaluated. You should reuse the old ones. Or I am missing something?

Well sure, but how do you supposed I get from 
`numbers: { a: 1, b: 2 }` to `numbers: [{ key: 'a', value: 1 }, { key: 'b', value: 2}]` without mapping if this is the response from the server? I'd have to do it somewhere.

Fair point about the array, seems like a good reason to keep that API as is, but doesn't prevent changing the object one.

**Comment by LiborOl** [comment]
You should create new array and merge there existing items from old array and insert new items from the response (and remove items that are not in the response anymore).
I don't know, if you have really only numbers(simple values) in the response object are there are some complex objects. If there are some complex objects, you should do the merge with old array anyway.

**Comment by HeyImAlex** [comment]
Yeah, I think adding an index to the array selectors would be a mistake. Arrays are actually treated more like sets internally, so to add indexes we could:

 - Make each element memoized by index (like maps and objects are memoized by keys). That would be terrible for performance as you've said; deleting the first element of an array would lead to every remaining element being recomputed.
- Use the same memoizing system but still pass the index. This makes the result of the selector impure, which would lead to confusing cache invalidation issues.
- Add a separate function for doing it this way. But I don't like having two separate nuanced ways to map over arrays. If it's really needed, the lower level `objectMemoize` makes it easy enough to write yourself.

That none of these options are particularly good is why I didn't include the key in the first place. However, that was kind of shortsighted because obviously it doesn't cause any issues with key-based memoization. Though it might be a little confusing that they're not perfectly symetric, I think that adding keys to key-based memoization functions and leaving set based functions alone is the way forward.

tl;dr: ya

**Comment by HeyImAlex** [comment]
Bikeshedding on how the key should be passed:

Mapping apis usually look like `(value, key, collection)`, and that's probably because `key` often isn't used and callbacks can just leave out that argument. I would imagine that the common case will be ignoring the key, so I think we should optimize for that case.

With `(value, key, ...args)`, you need to include the key argument even if you don't want it. If we were going to force the key, I would actually prefer `(key, value, ...args)` since I usually think of keys and values in that order. But right now I'm leaning toward the more clunky `(value, ...args, key)` that you suggested earlier. Key is optional, it *kind of* mirrors the mapping convention, and it's not breaking.

If there aren't any objections, I can write up this change pretty quick, or if someone else wants to submit a PR that's cool tool.

**Comment by mjrussell** [comment]
I agree that while `(value, ...args, key)` looks clunky at first, it does seem more similar to other libraries and is not breaking. Im for it, and happy to submit a PR if we are agreed. Should be a 1 line change and maybe a new test.

**Comment by HeyImAlex** [comment]
Alright, should be good to go. Thanks to everyone involved! Hopefully I didn't break anything. It's been a somber morning...
